diff --git a/out/resolve-dependency.js b/out/resolve-dependency.js
index 18a756b1a9f27f3571e2824810301eda4236fea6..dce9e7ef6dea9773185c30012283ce3ca91140f9 100644
--- a/out/resolve-dependency.js
+++ b/out/resolve-dependency.js
@@ -138,6 +138,40 @@ function getExportsTarget(exports, conditions, cjsResolve, moduleSyncAutoSelecta
     }
     return undefined;
 }
+function patternKeyCompare(a, b) {
+    const aPatternIndex = a.indexOf('*');
+    const bPatternIndex = b.indexOf('*');
+    const aBaseLength = aPatternIndex === -1 ? a.length : aPatternIndex + 1;
+    const bBaseLength = bPatternIndex === -1 ? b.length : bPatternIndex + 1;
+    if (aBaseLength > bBaseLength)
+        return -1;
+    if (bBaseLength > aBaseLength)
+        return 1;
+    if (aPatternIndex === -1)
+        return 1;
+    if (bPatternIndex === -1)
+        return -1;
+    if (a.length > b.length)
+        return -1;
+    if (b.length > a.length)
+        return 1;
+    return 0;
+}
+function getPatternMatch(pattern, subpath) {
+    const patternIndex = pattern.indexOf('*');
+    if (patternIndex === -1 || pattern.lastIndexOf('*') !== patternIndex) {
+        return undefined;
+    }
+    const patternBase = pattern.slice(0, patternIndex);
+    if (!subpath.startsWith(patternBase))
+        return undefined;
+    const patternTrailer = pattern.slice(patternIndex + 1);
+    if (subpath.length < pattern.length ||
+        (patternTrailer && !subpath.endsWith(patternTrailer))) {
+        return undefined;
+    }
+    return subpath.slice(patternIndex, subpath.length - patternTrailer.length);
+}
 function addExportsTargetPath(paths, pkgPath, target, wildcardReplacement) {
     if (typeof target !== 'string' || !target.startsWith('./'))
         return;
@@ -214,14 +248,13 @@ async function resolveExportsImports(pkgPath, obj, subpath, job, isImports, cjsR
             return Array.isArray(resolved) ? resolved : [resolved];
         }
     }
-    for (const match of Object.keys(matchObj).sort((a, b) => b.length - a.length)) {
-        if (match.endsWith('*') && subpath.startsWith(match.slice(0, -1))) {
+    for (const match of Object.keys(matchObj).sort(patternKeyCompare)) {
+        const wildcardReplacement = getPatternMatch(match, subpath);
+        if (wildcardReplacement !== undefined) {
             const target = getExportsTarget(matchObj[match], job.conditions, cjsResolve, nodeSupportsModuleSync);
             if (typeof target === 'string' && target.startsWith('./')) {
-                const resolvedPath = pkgPath +
-                    target.slice(1).replace(/\*/g, subpath.slice(match.length - 1));
+                const resolvedPath = pkgPath + target.slice(1).replace(/\*/g, wildcardReplacement);
                 const paths = [resolvedPath];
-                const wildcardReplacement = subpath.slice(match.length - 1);
                 const exportsForSubpath = matchObj[match];
                 if (job.moduleSyncCatchall &&
                     typeof exportsForSubpath === 'object' &&
